### PR TITLE
add mocks to allow cross os testing

### PR DESCRIPTION
**Pull request by tuxmea**
taken from
https://github.com/maestrodev/maestro-puppet-example/blob/master/spec/spec_helper.rb

work around https://tickets.puppetlabs.com/browse/PUP-1547
Ticket mentions fixed in 3.6.0 but rspec testing e.g. on OSX still
throws errors for package ensure set to latest and user with managehome
attribute and exec with user set to not root

should this be configurable?

### CODE REVIEW

**Comment by tuxmea** [comment]
Seems as if this also is a fix for
#387 #376 #256 #175

**Comment by daenney** [comment]
This looks sane to me. Can you add an example or two to further elaborate on what it solves? I know you linked some issues but for completeness it would be nice to see the use cases here.

File: Gemfile [hunk 1]
```diff
@@ -12,7 +12,12 @@ end
 
 gemspec
 
-gem 'rake'
+if RUBY_VERSION < '1.9.3'
```
Comment (bastelfreak) [inline Gemfile]: instead of this, could we drop ruby1.8 support? This is so old and EOL since a long time. voxpupuli started to drop ruby1.8 for each new module release a few months back.

Comment (tuxmea) [inline Gemfile]: My 5pc: There are still many Puppet 2.6, 2.7 on ruby 1.8 installations around the planet. Keeping at least rspec-puppet and puppet-lint with ruby 1.8 support will allow these people to easier migrate to newer versions of puppet. YMMV.
I suppose that dropping ruby 1.8 would require a major version bump.

Comment (rnelson0) [inline Gemfile]: Yes, dropping ruby 1.8 requires a new major version, and at least one minor version that announces the support will be deprecated.

As to whether we should... I'm not certain that using 1.8.7 will allow all the modern tooling to work anyway. I'd announce deprecation now and move to drop it shortly. Puppet 3 will be EOS in December, after all.

Comment (rnelson0) [inline Gemfile]: I also think that this deprecation should be in a separate commit rather than tied to mock changes.

File: lib/rspec-puppet/spec_helper.rb [hunk 2]
```diff
@@ -6,4 +6,19 @@
...
   c.environmentpath = File.join(Dir.pwd, 'spec')
+  c.before(:each) do
+    # work around https://tickets.puppetlabs.com/browse/PUP-1547
+    # ensure that there's at least one provider available by emulating that any
+    if Puppet::version < '3.2'
+      # ONLY WORKING WITH PUPPET < 3.2 !!
+      require 'puppet/provider/confine/exists'
+      Puppet::Provider::Confine::Exists.any_instance.stubs(:which => '')
+    else
+      # ONLY WORKING WITH PUPPET > 3.2 !!
```
Comment (rnelson0) [inline lib/rspec-puppet/spec_helper.rb]: `>= 3.2` right?

**Comment by rnelson0** [comment]
Ah, I see why you wanted to kill 1.8.7. I think in a separate PR, you should do that and mark 1.8.7 as allowable failures, then rebase this on top of it.

**Comment by tuxmea** [comment]
More descriptions:
when not (!) using RspecPuppetFacts

1. Package declaration:
```
  package { $::demo::package_name:
    ensure => latest,
  }
```
will result on OS X to:
```
  1) demo supported operating systems on Debian demo class without any parameters should compile into a catalogue without dependency cycles
     Failure/Error: it { is_expected.to compile.with_all_deps }
       error during compilation: Parameter ensure failed on Package[demo]: Provider must have features 'upgradeable' to set 'ensure' to 'latest' at /Users/mea/demo/spec/fixtures/modules/demo/manifests/install.pp:9
     # ./spec/classes/example_spec.rb:16
```
Test Code only sets osfamily fact:
```
      os = 'Debian'
      context "on #{os}" do
        let(:facts) do
          {
            :osfamily => 'Debian',
          }
        end

        context "demo class without any parameters" do
          it { is_expected.to compile.with_all_deps }
```

2. User declaration
```
  user { 'demo':
    ensure     => present,
    managehome => true,
    groups     => ['foo', 'bar'],
  }
```
will result on OS X to:
```
  1) demo supported operating systems on Debian demo class without any parameters should compile into a catalogue without dependency cycles
     Failure/Error: it { is_expected.to compile.with_all_deps }
       error during compilation: Parameter managehome failed on User[demo]: User provider directoryservice can not manage home directories at /Users/mea/demo/spec/fixtures/modules/demo/manifests/install.pp:16
     # ./spec/classes/example_spec.rb:16
```

**Comment by tuxmea** [comment]
ruby 1.8.7 is in #394

**Comment by binford2k** [comment]
If this eventually leads to me not needing crap like this:

```ruby
# fake out the file checks so that they validate as absolute even though they're Windows paths
before :each do
  Puppet[:autosign] = false
  Puppet::Util::Platform.stubs(:windows?).returns true
end
```

then I'll buy you many beers at puppetconf.

**Comment by tuxmea** [comment]
Code is in the wrong file. I am still trying to get this working directly in rspec-puppet.
